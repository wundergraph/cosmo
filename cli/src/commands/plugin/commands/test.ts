import { Command } from 'commander';
import { resolve } from 'pathe';
import pc from 'picocolors';
import Spinner from 'ora';
import { BaseCommandOptions } from '../../../core/types/types.js';
import { checkAndInstallTools, runGoTests } from '../toolchain.js';

export default (opts: BaseCommandOptions) => {
  const command = new Command('test');
  command.description('Run tests for a gRPC router plugin');
  command.argument('[directory]', 'Directory of the plugin', '.');
  command.option('-l, --language <language>', 'Programming language of the plugin', 'go');
  command.action(async (directory, options) => {
    const startTime = performance.now();
    const pluginDir = resolve(directory);
    const spinner = Spinner({ text: 'Running tests...' });

    try {
      spinner.start();

      if (options.language.toLowerCase() !== 'go') {
        spinner.warn(pc.yellow(`Language '${options.language}' is not supported yet. Using 'go' instead.`));
        options.language = 'go';
      }

      // Check and install tools if needed
      if (!options.skipToolsInstallation) {
        await checkAndInstallTools(options.forceToolsInstallation);
      }

      const srcDir = resolve(pluginDir, 'src');

      spinner.text = 'Running tests...';

      const { failed } = await runGoTests(srcDir, spinner);

      // Calculate elapsed time
      const endTime = performance.now();
      const elapsedTimeMs = endTime - startTime;
      const formattedTime =
        elapsedTimeMs > 1000 ? `${(elapsedTimeMs / 1000).toFixed(2)}s` : `${Math.round(elapsedTimeMs)}ms`;

      if (failed) {
        spinner.fail(pc.red(`Tests failed! [${formattedTime}]`));
      } else {
        spinner.succeed(pc.green(`Tests completed successfully! [${formattedTime}]`));
      }
    } catch (error: any) {
      spinner.fail(pc.red(`Failed to run tests: ${error.message}`));
      throw error;
    } finally {
      spinner.stop();
    }
  });

  return command;
};
