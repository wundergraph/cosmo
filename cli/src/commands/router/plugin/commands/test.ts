import path from 'node:path';
import os from 'node:os';
import { Command } from 'commander';
import { resolve } from 'pathe';
import Spinner from 'ora';
import { BaseCommandOptions } from '../../../../core/types/types.js';
import { checkAndInstallTools, installGoDependencies, runGoTests } from '../toolchain.js';
import { renderResultTree } from '../helper.js';

export default (opts: BaseCommandOptions) => {
  const command = new Command('test');
  command.description('Run tests for a gRPC router plugin');
  command.argument('[directory]', 'Directory of the plugin', '.');
  command.option('--skip-tools-installation', 'Skip tool installation', false);
  command.option(
    '--force-tools-installation',
    'Force tools installation regardless of version check or confirmation',
    false,
  );
  command.action(async (directory, options) => {
    const startTime = performance.now();
    const pluginDir = resolve(directory);
    const spinner = Spinner({ text: 'Running tests...' });
    const pluginName = path.basename(pluginDir);

    try {
      spinner.start();

      // Check and install tools if needed
      if (!options.skipToolsInstallation) {
        await checkAndInstallTools(options.forceToolsInstallation);
      }

      spinner.text = 'Installing Go dependencies...';

      await installGoDependencies(pluginDir, spinner);

      const srcDir = resolve(pluginDir, 'src');

      spinner.text = 'Running tests...';

      try {
        const { failed } = await runGoTests(srcDir, spinner, false);

        // Calculate elapsed time
        const endTime = performance.now();
        const elapsedTimeMs = endTime - startTime;
        const formattedTime =
          elapsedTimeMs > 1000 ? `${(elapsedTimeMs / 1000).toFixed(2)}s` : `${Math.round(elapsedTimeMs)}ms`;

        // Common details for both success and failure
        const details = {
          source: srcDir,
          env: `${os.platform()} ${os.arch()}`,
          time: formattedTime,
        };

        const title = failed ? 'Tests failed!' : 'Tests completed successfully!';
        renderResultTree(spinner, title, !failed, pluginName, details);
      } catch (error: any) {
        renderResultTree(spinner, 'Tests failed!', false, pluginName, {
          source: srcDir,
          env: `${os.platform()} ${os.arch()}`,
          error: error.message,
        });
      }
    } catch (error: any) {
      renderResultTree(spinner, 'Failed to run tests!', false, pluginName, {
        env: `${os.platform()} ${os.arch()}`,
        error: error.message,
      });
    } finally {
      spinner.stop();
    }
  });

  return command;
};
