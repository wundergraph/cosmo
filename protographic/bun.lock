{
  "lockfileVersion": 1,
  "workspaces": {
    "": {
      "name": "protographic",
      "dependencies": {
        "graphql": "^16.11.0",
        "lodash": "^4.17.21",
      },
      "devDependencies": {
        "@types/bun": "latest",
        "@types/lodash": "^4.17.16",
      },
      "peerDependencies": {
        "typescript": "^5",
      },
    },
  },
  "packages": {
    "@types/bun": ["@types/bun@1.2.11", "", { "dependencies": { "bun-types": "1.2.11" } }, "sha512-ZLbbI91EmmGwlWTRWuV6J19IUiUC5YQ3TCEuSHI3usIP75kuoA8/0PVF+LTrbEnVc8JIhpElWOxv1ocI1fJBbw=="],

    "@types/lodash": ["@types/lodash@4.17.16", "", {}, "sha512-HX7Em5NYQAXKW+1T+FiuG27NGwzJfCX3s1GjOa7ujxZa52kjJLOr4FUxT+giF6Tgxv1e+/czV/iTtBw27WTU9g=="],

    "@types/node": ["@types/node@22.15.3", "", { "dependencies": { "undici-types": "~6.21.0" } }, "sha512-lX7HFZeHf4QG/J7tBZqrCAXwz9J5RD56Y6MpP0eJkka8p+K0RY/yBTW7CYFJ4VGCclxqOLKmiGP5juQc6MKgcw=="],

    "bun-types": ["bun-types@1.2.11", "", { "dependencies": { "@types/node": "*" } }, "sha512-dbkp5Lo8HDrXkLrONm6bk+yiiYQSntvFUzQp0v3pzTAsXk6FtgVMjdQ+lzFNVAmQFUkPQZ3WMZqH5tTo+Dp/IA=="],

    "graphql": ["graphql@16.11.0", "", {}, "sha512-mS1lbMsxgQj6hge1XZ6p7GPhbrtFwUFYi3wRzXAC/FmYnyXMTvvI3td3rjmQ2u8ewXueaSvRPWaEcgVVOT9Jnw=="],

    "lodash": ["lodash@4.17.21", "", {}, "sha512-v2kDEe57lecTulaDIuNTPy3Ry4gLGJ6Z1O3vE1krgXZNrsQ+LFTGHVxVjcXPs17LhbZVGedAJv8XZ1tvj5FvSg=="],

    "typescript": ["typescript@5.8.3", "", { "bin": { "tsc": "bin/tsc", "tsserver": "bin/tsserver" } }, "sha512-p1diW6TqL9L07nNxvRMM7hMMw4c5XOo/1ibL4aAIGmSAt9slTE1Xgw5KWuof2uTOvCg9BY7ZRi+GaF+7sfgPeQ=="],

    "undici-types": ["undici-types@6.21.0", "", {}, "sha512-iwDZqg0QAGrg9Rav5H4n0M64c3mkR59cJ6wQp+7C4nI0gsmExaedaYLNO44eT4AtBBwjbTiGPMlt2Md0T9H9JQ=="],
  }
}
