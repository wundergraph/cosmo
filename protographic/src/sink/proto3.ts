import type { Service } from "@/common/service";
import { camelCase, snakeCase, upperFirst } from "lodash";
import type {
  ArgumentMapping,
  FieldMapping,
  OperationMapping,
  ServiceMapping,
  TypeFieldMapping,
} from "@/common/mapping";

// Our lightweight Proto AST types
interface ProtoAST {
  syntax: string;
  package: string;
  services: ProtoService[];
  messages: ProtoMessage[];
}

interface ProtoService {
  name: string;
  rpcs: ProtoRPC[];
}

interface ProtoRPC {
  name: string;
  requestType: string;
  responseType: string;
  isResponseStream?: boolean;
}

interface ProtoMessage {
  name: string;
  fields: ProtoField[];
}

interface ProtoField {
  name: string;
  type: string;
  number: number;
  isRepeated?: boolean;
  isOptional?: boolean;
  oneof?: ProtoOneOf;
}

interface ProtoOneOf {
  name: string;
  fields: ProtoField[];
}

export function intoProto3(svc: Service): {
  proto: ProtoAST;
  mapping: ServiceMapping;
} {
  const ast: ProtoAST = {
    syntax: "proto3",
    package: svc.name,
    services: [],
    messages: [],
  };

  const mapping: ServiceMapping = {
    version: 1,

    service: svc.name,
    operation_mappings: [],
    entity_mappings: [],
    type_field_mappings: [],
  };

  const protoService: ProtoService = {
    name: svc.name,
    rpcs: [],
  };

  const typeFieldMappingMap: Record<
    string,
    Omit<TypeFieldMapping, "type">
  > = {};

  // const typeFieldMapping: TypeFieldMapping = {
  //   type: rpc.kind as string,
  //   field_mappings: [],
  // };

  // RPCs and their request/response messages
  for (const rpc of svc.rpcs) {
    const protoRPCName = formatProtoRPCName(rpc.kind, rpc.name);
    const protoRPCRequest = formatProtoMessageName(`${protoRPCName}_Request`);
    const protoRPCResponse = formatProtoMessageName(`${protoRPCName}_Response`);

    const operationMapping: OperationMapping = {
      kind: rpc.kind,
      original: rpc.name,
      mapped: protoRPCName,
      request: protoRPCRequest,
      response: protoRPCResponse,
    };

    mapping.operation_mappings.push(operationMapping);

    const typeFieldMapping = typeFieldMappingMap[rpc.kind] || {
      field_mappings: [],
    };

    const fieldMapping: FieldMapping = {
      original: rpc.name,
      mapped: protoRPCName,
      argument_mappings: [],
    };

    typeFieldMapping.field_mappings.push(fieldMapping);

    const protoRequestMessage: ProtoMessage = {
      name: protoRPCRequest,
      fields: [],
    };

    for (let [index, argument] of rpc.arguments.entries()) {
      const protoFieldName = formatProtoMessageFieldName(argument.name);

      const argumentMapping: ArgumentMapping = {
        original: argument.name,
        mapped: protoFieldName,
      };

      fieldMapping.argument_mappings.push(argumentMapping);

      protoRequestMessage.fields.push({
        name: protoFieldName,
        type: mapTypeToProtoType(argument.type.name),
        isRepeated: argument.type.list,
        isOptional: !argument.type.required,

        number: index + 1,
      });
    }

    typeFieldMappingMap[rpc.kind] = typeFieldMapping;

    const protoResponseMessage: ProtoMessage = {
      name: protoRPCResponse,
      fields: [
        {
          name: formatProtoMessageFieldName(rpc.name),
          type: mapTypeToProtoType(rpc.type.name),
          isRepeated: rpc.type.list,
          isOptional: !rpc.type.required,

          number: 1,
        },
      ],
    };

    ast.messages.push(protoRequestMessage);
    ast.messages.push(protoResponseMessage);

    protoService.rpcs.push({
      name: protoRPCName,
      requestType: protoRequestMessage.name,
      responseType: protoResponseMessage.name,
    });
  }

  // Normal messages
  for (const message of svc.messages) {
    const protoMessage: ProtoMessage = {
      name: formatProtoMessageName(message.name),
      fields: message.fields.map((field, index) => ({
        name: formatProtoMessageFieldName(field.name),
        type: mapTypeToProtoType(field.type.name),

        isRepeated: field.type.list,
        isOptional: !field.type.required,

        number: index + 1,
      })),
    };

    ast.messages.push(protoMessage);
  }

  ast.services.push(protoService);

  for (const [type, fields] of Object.entries(typeFieldMappingMap)) {
    mapping.type_field_mappings.push({
      type,
      field_mappings: fields.field_mappings,
    });
  }

  console.log(JSON.stringify(mapping, undefined, 2));

  return { proto: ast, mapping: mapping };
}

export function printProto3(ast: ProtoAST): string {
  let output = "";

  // Print syntax and package
  output += `syntax = "${ast.syntax}";\n`;
  output += `package ${ast.package};\n\n`;

  // Print services
  for (const service of ast.services) {
    output += `service ${service.name} {\n`;
    for (const rpc of service.rpcs) {
      const responseType = rpc.isResponseStream
        ? `stream ${rpc.responseType}`
        : rpc.responseType;
      output += `  rpc ${rpc.name}(${rpc.requestType}) returns (${responseType}) {}\n`;
    }
    output += "}\n\n";
  }

  // Sort messages to match expected order
  const sortedMessages = [...ast.messages].sort((a, b) => {
    // First, group messages by their base name (without Request/Response/etc.)
    const aBase = a.name.replace(/(Request|Response|Input|Key|Result)$/, "");
    const bBase = b.name.replace(/(Request|Response|Input|Key|Result)$/, "");

    if (aBase !== bBase) {
      return aBase.localeCompare(bBase);
    }

    // Within the same base name, sort by suffix in specific order
    const suffixOrder = ["Request", "Input", "Key", "Response", "Result"];
    const aIndex = suffixOrder.findIndex((suffix) => a.name.endsWith(suffix));
    const bIndex = suffixOrder.findIndex((suffix) => b.name.endsWith(suffix));

    // If neither has a suffix, keep original order
    if (aIndex === -1 && bIndex === -1) {
      return 0;
    }

    // Put messages without suffix at the end
    if (aIndex === -1) return 1;
    if (bIndex === -1) return -1;

    return aIndex - bIndex;
  });

  // Print messages
  for (const message of sortedMessages) {
    output += `message ${message.name} {`;
    if (message.fields.length === 0) {
      output += "}\n\n";
      continue;
    }
    output += "\n";
    for (const field of message.fields) {
      if (field.oneof) {
        output += `  oneof ${field.oneof.name} {\n`;
        for (const oneofField of field.oneof.fields) {
          output += `    ${oneofField.type} ${oneofField.name} = ${oneofField.number};\n`;
        }
        output += "  }\n";
      } else {
        const repeated = field.isRepeated ? "repeated " : "";
        const optional = field.isOptional ? "optional " : "";
        output += `  ${repeated}${optional}${field.type} ${field.name} = ${field.number};\n`;
      }
    }
    output += "}\n\n";
  }

  return output.trim();
}

function mapTypeToProtoType(type: string): string {
  const typeMap: Record<string, string> = {
    ID: "string",
    String: "string",
    Int: "int32",
    Float: "float",
    Boolean: "bool",
  };
  return typeMap[type] || formatProtoMessageName(type);
}

function formatProtoMessageFieldName(name: string): string {
  return snakeCase(name);
}

function formatProtoMessageName(name: string): string {
  return upperFirst(camelCase(name));
}

function formatProtoRPCName(type: Operation, name: string): string {
  return `${upperFirst(type)}${formatProtoMessageName(name)}`;
}
