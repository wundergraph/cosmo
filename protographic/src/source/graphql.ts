import {
  Kind,
  parse,
  type DefinitionNode,
  type FieldDefinitionNode,
  type InputObjectTypeDefinitionNode,
  type InputValueDefinitionNode,
  type ObjectTypeDefinitionNode,
  type TypeNode,
} from "graphql";
import {
  type Message,
  type RPC,
  type Field,
  type Service,
  ServiceBuilder,
  type Type,
} from "@/common/service";

export function intoIntermediate(serviceName: string, sdl: string): Service {
  const ast = parse(sdl);

  const svcBuilder = new ServiceBuilder(serviceName);

  for (const node of ast.definitions) {
    switch (node.kind) {
      case Kind.OBJECT_TYPE_DEFINITION:
        switch (node.name.value) {
          case "Subscription":
          case "Mutation":
          case "Query":
            svcBuilder.addRPCs(...operationObjectToRPCs(node));
            break;
          default:
            svcBuilder.addMessages(objectToMessage(node));
            break;
        }

        break;
      case Kind.INPUT_OBJECT_TYPE_DEFINITION:
        svcBuilder.addMessages(objectToMessage(node));
        break;
      default:
        console.error("unsupported graphql type used: " + node.kind);
    }
  }

  return svcBuilder.build();
}

function operationObjectToRPCs(node: ObjectTypeDefinitionNode): RPC[] {
  const rpcs: RPC[] = [];

  for (const field of node.fields ?? []) {
    rpcs.push(operationFieldToRPC(node.name.value, field));
  }

  return rpcs;
}

function operationFieldToRPC(type: string, node: FieldDefinitionNode): RPC {
  const rpc: RPC = {
    name: node.name.value,
    kind: "Query",
    type: typeFromTypeNode(node.type),
    arguments: [],
  };

  switch (type) {
    case "Subscription":
      rpc.kind = "Subscription";
      break;
    case "Mutation":
      rpc.kind = "Mutation";
      break;
    case "Query":
      rpc.kind = "Query";
      break;
    default:
      console.error("unsupported graphql type used: " + type);
  }

  for (const field of node.arguments ?? []) {
    rpc.arguments.push(fieldNodeToField(field));
  }

  return rpc;
}

function fieldNodeToField(
  input: InputValueDefinitionNode | FieldDefinitionNode,
): Field {
  const field: Field = {
    name: input.name.value,
    type: typeFromTypeNode(input.type),

    resolved: false,
  };

  return field;
}

function typeFromTypeNode(type: TypeNode): Type {
  const typeObj: Type = { name: "", required: false, list: false };

  let fieldType = type;

  if (fieldType.kind === Kind.NON_NULL_TYPE) {
    typeObj.required = true;
    fieldType = fieldType.type;
  }

  if (fieldType.kind === Kind.LIST_TYPE) {
    typeObj.list = true;
    if (fieldType.type.kind === Kind.NON_NULL_TYPE) {
      fieldType = fieldType.type.type;
    } else {
      fieldType = fieldType.type;
    }
  }

  if (fieldType.kind === Kind.NAMED_TYPE) {
    typeObj.name = fieldType.name.value;
  }

  return typeObj;
}

function objectToMessage(
  node: ObjectTypeDefinitionNode | InputObjectTypeDefinitionNode,
): Message {
  const message: Message = {
    name: node.name.value,
    fields: [],
  };

  for (const field of node.fields ?? []) {
    message.fields.push(fieldNodeToField(field));
  }

  return message;
}
