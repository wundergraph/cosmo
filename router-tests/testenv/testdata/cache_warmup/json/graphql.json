{
  "query": "query { employees { id details { forename } } }"
}