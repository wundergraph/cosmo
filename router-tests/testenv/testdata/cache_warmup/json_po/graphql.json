{
  "request": {
    "extensions": {
      "persistedQuery": {
        "version": 1,
        "sha256Hash": "dc67510fb4289672bea757e862d6b00e83db5d3cbbcfb15260601b6f29bb2b8f"
      }
    }
  },
  "client": {
    "name": "my-client",
    "version": "missing"
  }
}