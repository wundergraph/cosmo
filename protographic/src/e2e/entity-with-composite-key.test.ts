import { describe, it, expect } from "bun:test";
// import {
//   validateProtoDefinition,
//   normalizeWhitespace,
// } from "../utils/proto-comparator";
import { intoIntermediate } from "@/source/graphql";
import { intoProto3, printProto3 } from "@/sink/proto3";

describe("Entity with Composite Key", () => {
  const sdl = `
    type Order @key(fields: "customerId orderId") {
      customerId: ID!
      orderId: ID!
      total: Float!
      items: [OrderItem!]!
    }

    type OrderItem {
      productId: ID!
      quantity: Int!
      price: Float!
    }
  `;

  it("should correctly read GraphQL SDL into intermediate format", () => {
    const result = intoIntermediate("OrderService", sdl);

    expect(result).toMatchSnapshot();
  });

  it("should generate correct proto3 ast", () => {
    const result = intoIntermediate("OrderService", sdl);
    const proto = intoProto3(result);

    expect(proto).toMatchSnapshot();
  });

  it("should generate correct proto3 schema", () => {
    const result = intoIntermediate("OrderService", sdl);
    const { proto } = intoProto3(result);
    const protoSchema = printProto3(proto);

    expect(protoSchema).toMatchSnapshot();
  });

  const expectedProto = `
    syntax = "proto3";

    package order;

    service OrderService {
      rpc LookupOrderByCustomerIdAndOrderId(LookupOrderByCustomerIdAndOrderIdRequest) returns (LookupOrderByCustomerIdAndOrderIdResponse) {}
    }

    message LookupOrderByCustomerIdAndOrderIdRequest {
      repeated LookupOrderByCustomerIdAndOrderIdInput inputs = 1;
    }

    message LookupOrderByCustomerIdAndOrderIdInput {
      OrderByCustomerIdAndOrderIdKey key = 1;
    }

    message LookupOrderByCustomerIdAndOrderIdResponse {
      repeated LookupOrderByCustomerIdAndOrderIdResult results = 1;
    }

    message LookupOrderByCustomerIdAndOrderIdResult {
      Order order = 1;
    }

    message Order {
      string customerId = 1;
      string orderId = 2;
      float total = 3;
      repeated OrderItem items = 4;
    }

    message OrderByCustomerIdAndOrderIdKey {
      string customerId = 1;
      string orderId = 2;
    }

    message OrderItem {
      string productId = 1;
      int32 quantity = 2;
      float price = 3;
    }
  `;
});
