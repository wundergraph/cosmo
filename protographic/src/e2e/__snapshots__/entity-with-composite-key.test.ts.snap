// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Entity with Composite Key should correctly read GraphQL SDL into intermediate format 1`] = `
{
  "messages": [
    {
      "fields": [
        {
          "name": "customerId",
          "resolved": false,
          "type": {
            "list": false,
            "name": "ID",
            "required": true,
          },
        },
        {
          "name": "orderId",
          "resolved": false,
          "type": {
            "list": false,
            "name": "ID",
            "required": true,
          },
        },
        {
          "name": "total",
          "resolved": false,
          "type": {
            "list": false,
            "name": "Float",
            "required": true,
          },
        },
        {
          "name": "items",
          "resolved": false,
          "type": {
            "list": true,
            "name": "OrderItem",
            "required": true,
          },
        },
      ],
      "name": "Order",
    },
    {
      "fields": [
        {
          "name": "productId",
          "resolved": false,
          "type": {
            "list": false,
            "name": "ID",
            "required": true,
          },
        },
        {
          "name": "quantity",
          "resolved": false,
          "type": {
            "list": false,
            "name": "Int",
            "required": true,
          },
        },
        {
          "name": "price",
          "resolved": false,
          "type": {
            "list": false,
            "name": "Float",
            "required": true,
          },
        },
      ],
      "name": "OrderItem",
    },
  ],
  "name": "OrderService",
  "rpcs": [],
}
`;

exports[`Entity with Composite Key should generate correct proto3 ast 1`] = `
{
  "mapping": {
    "entity_mappings": [],
    "operation_mappings": [],
    "service": "OrderService",
    "type_field_mappings": [],
    "version": 1,
  },
  "proto": {
    "messages": [
      {
        "fields": [
          {
            "isOptional": false,
            "isRepeated": false,
            "name": "customer_id",
            "number": 1,
            "type": "string",
          },
          {
            "isOptional": false,
            "isRepeated": false,
            "name": "order_id",
            "number": 2,
            "type": "string",
          },
          {
            "isOptional": false,
            "isRepeated": false,
            "name": "total",
            "number": 3,
            "type": "float",
          },
          {
            "isOptional": false,
            "isRepeated": true,
            "name": "items",
            "number": 4,
            "type": "OrderItem",
          },
        ],
        "name": "Order",
      },
      {
        "fields": [
          {
            "isOptional": false,
            "isRepeated": false,
            "name": "product_id",
            "number": 1,
            "type": "string",
          },
          {
            "isOptional": false,
            "isRepeated": false,
            "name": "quantity",
            "number": 2,
            "type": "int32",
          },
          {
            "isOptional": false,
            "isRepeated": false,
            "name": "price",
            "number": 3,
            "type": "float",
          },
        ],
        "name": "OrderItem",
      },
    ],
    "package": "OrderService",
    "services": [
      {
        "name": "OrderService",
        "rpcs": [],
      },
    ],
    "syntax": "proto3",
  },
}
`;

exports[`Entity with Composite Key should generate correct proto3 schema 1`] = `
"syntax = "proto3";
package OrderService;

service OrderService {
}

message Order {
  string customer_id = 1;
  string order_id = 2;
  float total = 3;
  repeated OrderItem items = 4;
}

message OrderItem {
  string product_id = 1;
  int32 quantity = 2;
  float price = 3;
}"
`;
